cache/disk/casblob/casblob.go

```diff
@@ -0,0 +1,597 @@
+
+	// With 1M chunks this gives a limit of ~511TB which is plenty.
+	//frameSize   uint32 // 4 bytes: the size of the rest of the header.
+
+	// Then we put our metadata:
+
+	uncompressedSize int64           // 8 bytes
+	compression      CompressionType // uint8, 1 byte
+	chunkSize        uint32          // 4 bytes
+
+	// Offsets in the file on disk, of each chunk, with an additional
+	// entry for the end of the file.
+	//
+	// Stored as an int64 number of chunks, followed by their int64 offsets,
+	// and a final value for the size of the file (header + data).
+	// 8 bytes + (n+1)*8 bytes.
+	chunkOffsets []int64
+}
+
+const chunkTableOffset = 4 + 4 + 8 + 1 + 4 + 8
+
+// Returns the size of the header itself.
+func (h *header) size() int64 {
+	return chunkTableOffset + (int64(len(h.chunkOffsets)) * 8)
+}
+
+func (h *header) frameSize() uint32 {
+	return chunkTableOffset + (uint32(len(h.chunkOffsets)) * 8) - 4 - 4
+}
+
+// Provides an io.ReadCloser that returns uncompressed data from a cas blob.
+type readCloserWrapper struct {
+	*header
+
+	rdr io.Reader // Read from this, not from decoder or file.
+
+	decoder *syncpool.DecoderWrapper // Might be nil.
+	file    *os.File
+}
+
+var errWrongMagicNum = errors.New("expected magic number not found")
+
+// Read the header and leave f at the start of the data.
+func readHeader(f *os.File) (*header, error) {
+	var err error
+	var h header
+
+	fileInfo, err := f.Stat()
+	if err != nil {
+		return nil, err
+	}
+	foundFileSize := fileInfo.Size()
+
+	var magicNumber uint32
+	err = binary.Read(f, binary.LittleEndian, &magicNumber)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read magic number: %w", err)
+	}
+	if magicNumber != skippableFrameMagicNumber {
+		return nil, errWrongMagicNum
+	}
+
+	var frameSize uint32
+	err = binary.Read(f, binary.LittleEndian, &frameSize)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read frameSize: %w", err)
+	}
+
+	err = binary.Read(f, binary.LittleEndian, &h.uncompressedSize)
+	if err != nil {
+		return nil, err
+	}
+
+	err = binary.Read(f, binary.LittleEndian, &h.compression)
+	if err != nil {
+		return nil, err
+	}
+
+	err = binary.Read(f, binary.LittleEndian, &h.chunkSize)
+	if err != nil {
+		return nil, err
+	}
+
+	var numOffsets int64
+	err = binary.Read(f, binary.LittleEndian, &numOffsets)
+	if err != nil {
+		return nil, err
+	}
+
+	if numOffsets < 2 {
+		// chunkOffsets has an extra entry to specify the compressed file size.
+		return nil, fmt.Errorf("internal error: need at least one chunk, found %d", numOffsets-1)
+	}
+
+	metadataSize := numOffsets*8 + 8 + 1 + 4 + 8
+	if int64(frameSize) != metadataSize {
+		return nil, fmt.Errorf("metadata frame size %d, but metadata size %d",
+			frameSize, metadataSize)
+	}
+
+	prevOffset := int64(-1)
+	h.chunkOffsets = make([]int64, numOffsets, numOffsets)
+	for i := 0; int64(i) < numOffsets; i++ {
+		err = binary.Read(f, binary.LittleEndian, &h.chunkOffsets[i])
```

Comment:
I don't understand the question... In the previous read we found the number of entries in the offset table, and here we create a slice of that size and add it to the header struct. Then we read into that slice. Are you asking it we could do a single Read call to populate the slice?

Reply:
[binary.Read ](https://golang.org/pkg/encoding/binary/#Read) seems to be able to read whole slices at once. Yeah, but cannot read a whole struct b/c of the slice inside so leaving up to you whether it would be clearer.

Reply:
Pushed a commit to read/write the slice in a single Read/Write call.